File: wi-seong-cheol/PGS/0x05 Stack/기능개발.cpp
//
//  기능개발.cpp
//  main
//
//  Created by wi_seong on 2022/11/09.
//

// 시간복잡도: O(n)
// 최악시간: 100
// 난이도: Level 2
// Timer: 12m
// Url: https://school.programmers.co.kr/learn/courses/30/lessons/42586
#include <string>
#include <vector>
#include <stack>
#include <cmath>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    stack<int> S;
    
    for(size_t i = 0; i < progresses.size(); i++) {
        int d = ceil((100 - progresses[i]) / double(speeds[i]));
        while(!S.empty() && S.top() < d) {
            S.pop();
        }
        if(S.empty()) answer.push_back(1);
        else answer.back()++;
        S.push(d);
    }
    
    return answer;
}

File: jokercsi/Programmers/기능개발.py
def solution(progresses, speeds):
    answer = []
    
    while True:
        for i in range(len(progresses)):
            progresses[i] += speeds[i]
            
        count = 0
        if progresses[0] >= 100:
            while progresses[0] >= 100:
                progresses.pop(0)
                speeds.pop(0)
                count += 1
                if not progresses:
                    break
            answer.append(count)
        
        if not progresses:
            break
        
    
    return answer
